### PR TITLE
feat: rebase anon poll joining with dev

### PR DESCRIPTION
**Pull request by 0xmad**
# Description

- [x] Add rebase missing parts
- [x] Remove sha256 input hash

## Additional Notes

Will add later as separate PRs:
- Update docs and circuits readme

## Related issue(s)

Related to #1838 

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: packages/circuits/circom/core/non-qv/processMessages.circom [hunk 1]
```diff
@@ -358,14 +338,13 @@ template ProcessOneNonQv(stateTreeDepth, voteOptionTreeDepth) {
...
-    (computedNewSlPubKey, computedNewBallotNonce, computedIsValid, computedIsStateLeafIndexValid, computedIsVoteOptionIndexValid) = StateLeafAndBallotTransformerNonQv()(
+    var computedNewSlPubKey[2], computedNewBallotNonce, computedIsValid;
+    (computedNewSlPubKey, computedNewBallotNonce, computedIsValid) = StateLeafAndBallotTransformerNonQv()(
         numSignUps,
         maxVoteOptions,
         [stateLeaf[STATE_LEAF_PUB_X_IDX], stateLeaf[STATE_LEAF_PUB_Y_IDX]],
         stateLeaf[STATE_LEAF_VOICE_CREDIT_BALANCE_IDX],
-        stateLeaf[STATE_LEAF_TIMESTAMP_IDX],
-        pollEndTimestamp,
+        // stateLeaf[STATE_LEAF_TIMESTAMP_IDX],
```
Comment (kittybest) [inline packages/circuits/circom/core/non-qv/processMessages.circom:347]: why this is commented out?

File: packages/circuits/circom/core/non-qv/processMessages.circom [hunk 2]
```diff
@@ -215,7 +195,7 @@ include "../../trees/incrementalQuinaryTree.circom";
             computedCommandsSigR8[i],
             computedCommandsSigS[i],
             computedCommandsPackedCommandOut[i]
-        ) = MessageToCommand()(msgs[i], coordPrivKey, encPubKeys[i]);
+        ) = MessageToCommand()(message, coordPrivKey, encPubKeys[i]);
```
Comment (ctrlc03) [inline packages/circuits/circom/core/non-qv/processMessages.circom:198]: why not pass msgs[i] as before?

File: packages/circuits/circom/utils/non-qv/stateLeafAndBallotTransformer.circom [hunk 3]
```diff
@@ -66,13 +62,9 @@ template StateLeafAndBallotTransformerNonQv() {
     
     // True when the command is valid; otherwise false.
     signal output isValid;
-    // True if the state leaf index is valid
```
Comment (ctrlc03) [inline packages/circuits/circom/utils/non-qv/stateLeafAndBallotTransformer.circom:69]: why reverting this?

**Review by ctrlc03 — CHANGES_REQUESTED** [review]
Thanks. Please take a look again at this and ensure that we are not reverting features/fixes/refactoring introduced since the feat/anon-poll-joining branch was merged. For instance:

* circuit templates returning certain output signals to prevent double checks
* max vote option 
* MACI.deployPoll storing all contracts (poll, mp, tally) rather than just poll

File: packages/circuits/circom/core/qv/pollJoining.circom [hunk 4]
```diff
@@ -0,0 +1,78 @@
+pragma circom 2.0.0;
+
+
+// circomlib import
+include "./mux1.circom";
```
[the remark this answers is not in the document]
Comment (0xmad) [reply packages/circuits/circom/core/qv/pollJoining.circom:5]: So, we don't need to implement it for non-qv?

File: packages/contracts/contracts/MACI.sol [hunk 5]
```diff
@@ -160,26 +169,32 @@ contract MACI is IMACI, DomainObjs, Params, Utilities {
...
   function deployPoll(
     uint256 _duration,
     TreeDepths memory _treeDepths,
+    uint8 _messageBatchSize,
     PubKey memory _coordinatorPubKey,
     address _verifier,
     address _vkRegistry,
     Mode _mode
-  ) public virtual {
+  ) public virtual returns (PollContracts memory pollAddr) {
```
[the remark this answers is not in the document]
Comment (0xmad) [reply packages/contracts/contracts/MACI.sol:197]: The same as https://github.com/privacy-scaling-explorations/maci/pull/1944#discussion_r1866074409

File: packages/circuits/circom/utils/non-qv/stateLeafAndBallotTransformer.circom (hunk 3 above)
Comment (0xmad) [reply packages/circuits/circom/utils/non-qv/stateLeafAndBallotTransformer.circom:69]: I thought it was done on purpose

File: packages/circuits/circom/core/non-qv/processMessages.circom (hunk 1 above)
Comment (0xmad) [reply packages/circuits/circom/core/non-qv/processMessages.circom:347]: This is an original change from anon joining branch

File: packages/circuits/circom/core/non-qv/processMessages.circom [hunk 6]
```diff
@@ -50,14 +44,14 @@ include "../../trees/incrementalQuinaryTree.circom";
...
     // Number of users that have completed the sign up.
     signal input numSignUps;
-    // Time when the poll ends.
-    signal input pollEndTimestamp;
-    // The existing message tree root.
-    signal input msgRoot;
+    // Number of options for this poll.
+    signal input maxVoteOptions;
+    // Value of chainHash at beginning of batch
+    signal input inputBatchHash;
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply packages/circuits/circom/core/non-qv/processMessages.circom:50]: probably no need actually as we have the output as public input 🤔

File: packages/contracts/contracts/PollFactory.sol [hunk 7]
```diff
@@ -20,23 +17,13 @@ contract PollFactory is Params, DomainObjs, IPollFactory {
   /// @inheritdoc IPollFactory
   function deploy(
     uint256 _duration,
-    TreeDepths calldata _treeDepths,
-    PubKey calldata _coordinatorPubKey,
-    address _maci,
-    uint256 _emptyBallotRoot
```
Comment (ctrlc03) [inline packages/contracts/contracts/PollFactory.sol:26]: let's put this back

**Review by ctrlc03 — CHANGES_REQUESTED** [review]
Thanks, added some more comments

File: packages/contracts/contracts/PollFactory.sol (hunk 7 above)
Comment (0xmad) [reply packages/contracts/contracts/PollFactory.sol:26]: Do you mean `_emptyBallotRoot`?

Comment (ctrlc03) [reply packages/contracts/contracts/PollFactory.sol:26]: yes not sure why it highlighted everything

File: packages/circuits/circom/core/non-qv/processMessages.circom [hunk 8]
```diff
@@ -412,8 +379,10 @@ template ProcessOneNonQv(stateTreeDepth, voteOptionTreeDepth) {
 
     // 5. Verify that currentVoteWeight exists in the ballot's vote option root
     // at cmdVoteOptionIndex.
-    b <== currentVoteWeight;
-    c <== cmdNewVoteWeight;
+    var voiceCreditAmountValid = SafeGreaterEqThan(252)([
```
Comment (ctrlc03) [inline packages/circuits/circom/core/non-qv/processMessages.circom:382]: checking on dev, this is not there and seems like it's not needed, we do the check already inside message valiadtor, if isValid is true then it means that the voice credits amount is valid

File: packages/circuits/circom/core/non-qv/processMessages.circom [hunk 9]
```diff
@@ -427,10 +396,17 @@ template ProcessOneNonQv(stateTreeDepth, voteOptionTreeDepth) {
     computedCurrentVoteWeightQip === ballot[BALLOT_VO_ROOT_IDX];
 
     var voteWeightMux = Mux1()([currentVoteWeight, cmdNewVoteWeight], computedIsValid);
+    var newSlVoiceCreditBalanceMux = Mux1()(
```
Comment (ctrlc03) [inline packages/circuits/circom/core/non-qv/processMessages.circom:399]: as above, this is not needed

File: packages/circuits/circom/core/non-qv/processMessages.circom [hunk 10]
```diff
@@ -427,10 +396,17 @@ template ProcessOneNonQv(stateTreeDepth, voteOptionTreeDepth) {
...
+        [
+            stateLeaf[STATE_LEAF_VOICE_CREDIT_BALANCE_IDX],
+            stateLeaf[STATE_LEAF_VOICE_CREDIT_BALANCE_IDX] + currentVoteWeight - cmdNewVoteWeight
+        ],
+        voiceCreditAmountValid
+    );
     var voiceCreditBalanceMux = Mux1()(
         [
             stateLeaf[STATE_LEAF_VOICE_CREDIT_BALANCE_IDX],
-            stateLeaf[STATE_LEAF_VOICE_CREDIT_BALANCE_IDX] + b - c
```
Comment (ctrlc03) [inline packages/circuits/circom/core/non-qv/processMessages.circom:433]: let's keep as is as it was an optimisation we added after this grant work

File: packages/circuits/circom/core/qv/processMessages.circom [hunk 11]
```diff
@@ -409,8 +379,13 @@ template ProcessOne(stateTreeDepth, voteOptionTreeDepth) {
 
     // 5. Verify that currentVoteWeight exists in the ballot's vote option root
     // at cmdVoteOptionIndex.
-    b <== currentVoteWeight * currentVoteWeight;
-    c <== cmdNewVoteWeight * cmdNewVoteWeight;
+    currentVoteWeightSquare <== currentVoteWeight * currentVoteWeight;
+    cmdNewVoteWeightSquare <== cmdNewVoteWeight * cmdNewVoteWeight;
+
+    var voiceCreditAmountValid = SafeGreaterEqThan(252)([
```
Comment (ctrlc03) [inline packages/circuits/circom/core/qv/processMessages.circom:385]: let's remove this (and the next part) as mentioned in the non-qv part

File: packages/contracts/contracts/Poll.sol [hunk 12]
```diff
@@ -201,28 +271,110 @@ contract Poll is Params, Utilities, SnarkCommon, IPoll {
...
     stateMerged = true;
 
-    uint256 _mergedStateRoot = extContracts.maci.getStateTreeRoot();
+    uint256 _mergedStateRoot = InternalLazyIMT._root(pollStateTree);
     mergedStateRoot = _mergedStateRoot;
 
     // Set currentSbCommitment
     uint256[3] memory sb;
-    sb[0] = _mergedStateRoot;
+    sb[0] = mergedStateRoot;
```
Comment (ctrlc03) [inline packages/contracts/contracts/Poll.sol:370]: this should be _mergedStateRoot

**Review by ctrlc03 — APPROVED** [review]
Thank you!

**Pull request merged**